### PR TITLE
Flutter Embeded V2

### PR DESCRIPTION
**Pull request by Larpoux**
This pull request address : #203 and #205 

The following tests had been done : 

| Device | SDK  | V1 Result  | V2 Result |  Comment |
| :------------ |:---------------:| :---------------:| :----------| :----------| 
| Android Emulator | SDK 16 | NOK | NOK | Does not start. Could be a problem with the emulator |
|                                 | SDK 17 | NOK   | NOK | Does not start. Could be a problem with the emulator |
|                                 | SDK 18 | OK .  | OK .  | Record is distorted on the Android Emulator |
|                                 | SDK 21 | OK .  | OK    | Record is distorted on the Android Emulator |
|                                 | SDK 22 | OK    | OK .  | Record is distorted on the Android Emulator |
|                                 | SDK 26 | OK    | OK .  | - |
|                                 | SDK 28 | OK    | OK .  | - |
|                                 | SDK 29 | OK    | OK .  | - |
| Real device : Xiaomi Mi Max 3| Android 8.1.0 | OK | OK| - |
| Real device : Huawei P20 | Android 9 | OK | OK| - |
| Real device : Huawei Mediapad M5 | Android 9 | OK | OK| - |
| iOS emulator | iOS 13.3 | OK | OK | - |
|              | iOS 12.4 | OK | OK | - |
|              | iOS 12.0 | OK | OK | - |
|              | iOS 11.4 | OK | OK | - |
|              | iOS 10.3 | OK | OK | - |
| Real device iPhone 5s | iOS 12.4.4 | OK | OK | - |
| Real device iPhone 11 Pro Max | iOS 13.3 | OK | OK | - |
| Real device iPad Pro | iOS 12.4 (jailbroken) | OK | OK | - |

### CODE REVIEW

File: android/src/main/AndroidManifest.xml [hunk 1]
```diff
@@ -1,3 +1,3 @@
 <manifest xmlns:android="http://schemas.android.com/apk/res/android"
-  package="com.dooboolab.fluttersound">
+   package="com.dooboolab.fluttersound">
```
Comment (hyochan) [inline android/src/main/AndroidManifest.xml]: 2 spaces to 3 spaces?

**Comment by hyochan** [comment]
I'll look into this after you've resolve the conflict with the master branch!

File: android/src/main/AndroidManifest.xml (hunk 1 above)
Comment (Larpoux) [reply android/src/main/AndroidManifest.xml]: 3 spaces is not good. I am going to fix that

**Comment by Larpoux** [comment]
OK Hyo, I think everything ready for my android-x branch

**Comment by bsutton** [comment]
@larpoux did you see my pull request that contains some important amendments to the pubspec.yaml files.
Also why has your pull requested modified 120 files including binary files?

**Comment by Larpoux** [comment]
> @Larpoux did you see my pull request that contains some important amendments to the pubspec.yaml files.

No, I did not. Where ?


> Also why has your pull requested modified 120 files including binary files?

No :-D :-D :-D
I did not modified 120 files. I do not have any time to do something like that. :-D
I just duplicated the ```example``` directory to ```example-emb-v2``` so that I could test my fork with ```emb V1``` **AND** ```emb V2```. I really wanted to be backward compatible, without any regression and I did many many tests to check that.

I did not try to keep a common trunk between ```example V1``` and ```example V2``` because they are just test apps, and probably ```example V1``` will be removed in the future.

If there is binary files in ```example V2``` this is probably normal. If not, this is because the .gitignore is not correct. In .gitignore, I just added the same entries than ```example V1``` and prefixed those entries by ```example-emb-V2```

**Comment by bsutton** [comment]
You will need to modify your pull request to remove the extra files.

As to my Pull Request (PR) you can see it here:
https://github.com/Canardoux/flutter_sound/pull/2

**Comment by bsutton** [comment]
You can see the list of files you have changed here:

https://github.com/dooboolab/flutter_sound/pull/207/files

**Comment by Larpoux** [comment]
@bsutton , you are not fair with me. You posted that I want to change 120 files.
But if you removed the two examples (which are not part of flutter_sound), there is only 9 files modified. I spent one minute of my life just to verify this.

In these 9 files you can group them in 3 categories : 
- Cosmetic changes (Hyo di not like my ```fin``` and ```fou```t variables and I changed them to ```inputFile``` and ```outputFile``` (5 lines changed)
- Some changes in the architecture (checking permission in the java side was really bad, was erroneous, and gave problems with V2 compatibility. (12 lines changed)
- And integration of Flutter embedded V2.

You must understand that my work on flutter_sound is not fun. Not fun at all. I love to code. But my work on flutter_sound is not coding. It is just integration. Integration of ffmpeg. Integration of Flutter Embedded V2. And now I do Quality Control with @salvatore373 .
I do this job because I know that this job must be done. By me or by somebody else.

If you do not like the two examples, just do a ```rm -r example example-emb-V2``` then,  you will have only 12 lines changed. But at the same time you will be naked to check if my work is correct and compatible with V1.

Please realise that I did not code anything new. Just copy and past the Google documentation. I was even not able to understand what I did. Just follow the documentation. This is really a stupid job. Programming is a stupid job : 98% is just garbage. Just 2% is really fun. But I will never forget the 98% : the 2% are so fun :-D.

Sorry if my English is not clear, I do not ride kangaroos ;-)

**Comment by bsutton** [comment]
@larpoux is not about bring fair, I'm trying to help you get you changes
merged.

Most developer will not accept a pull request with a large set of changes.
If you want to get a pr accepted every change most be justified.

I apologise if I came across as being unfair that was not my intention,
it's just that riding kangaroos all day makes me grumpy. It's all that
bouncing.

**Comment by bsutton** [comment]
Programming isn't a stupid job, we are artist creating beautiful things.

**Comment by Larpoux** [comment]
Thank you Brett. I appreciate your help. This PR was painful for me. Simple, very simple (just copy and past from Google documentation) but I had to to manage test with two OS, with two versions (V1 and V2) and with many versions. Many hours.

I think you have another problem than kangaroos : usually you walk on ceiling instead of the floor.
Probably hard for the brain.

**Comment by Larpoux** [comment]
>Programming isn't a stupid job, we are artist creating beautiful things.

I totally agree with you, we are artists. We can create beautiful things, but we can also produce just garbage. During my programer life I had to fight with a stupid boss, a stupid chief, a stupid little-chief, stupid clients. I had also to fight with my own laziness and my own stupidity.
This is also our programmer's life

**Comment by Larpoux** [comment]
I hope that fires in Australia are finished. I felt very concerned by this catastrophe

**Comment by bsutton** [comment]
We are only about 2/3 of the way through the fire season.

I'm not directly affected except we have had a few days where the haze from
the smoke is terrible.

**Comment by salvatore373** [comment]
I think that we should come back to coding 😂
Anyway, I'm sad too for the Australian catastrophe 🙏🏻

**Comment by Larpoux** [comment]
@bsutton : life in Australia is hard. Not only you usually walk on ceiling instead of the floor, and this is not comfortable, but because of the land attraction, everything have a tendency to fall up to the ceiling. Furnitures, TV, computer ...
There is probably a terrible mess on your ceiling, and you have to clean everything every Sundays.
And when you go to work and park your kangaroo  in the kangaroo-parking, you must attach him firmly so that he does fly away like a balloon.
I definitely will not like to live in a country located on the Antipodes.

**Comment by Larpoux** [comment]
@hyochan : could you kindly tell me what is the status of #207 ?
Do you plan to integrate it in Master Branch or do you reject it ?
Thank you. Larpoux

**Comment by hyochan** [comment]
@Larpoux Hi, my friend. Sorry for the delay. I am suffering from busy work these days. I am Korean and we do work a lot in our country. We always over work and we have less sleep. What's worst is that we don't gain much achievement from that work.

Sorry, I will come back to our thread again. I think the `PR` should be considered merging after the massive work given in #202. Since @salvatore373 have been taking time in resolving conflicts a few times, it'd be better to have that work in our `master` before we combine any other work. How do you think?

Hope we can finish #202 up first.

**Comment by Larpoux** [comment]
@hyochan , you are my friend too, Hyo 😺 

I understand that you have many things to do. And flutter_sound is not all your life, of course.
When I worked (for money), I was not able to work on free projects because I did not have enough time. I had to work for my employer, and I had to manage my family.
But now I am retired. I don't have to work for money any more. And I do not have to take care of my children anymore. So I have plenty of time to work on free project : my own App, and now a little bit for flutter_sound. Life is easy for me : I get money but I do not have to work for a private company.
You really must not be sorry. Your project is great and we need you as flutter_sound maintener.

Yes, I agree that integrate @salvatore373 fork (when he will finishes) is OK.
I think he did a very good job, and I will check his fork before you can integrate it in the Master Branch.

**Comment by salvatore373** [comment]
Thank you for your comprehension. I am doing my best to try to finish my work as fast as possible.

**Comment by Larpoux** [comment]
This pull request will be superseded by flauto

**Comment by bsutton** [comment]
So what is flauto?

**Comment by Larpoux** [comment]
The next version of flutter_sound will (eventually) have a new module allowing the user to control his sound app from the lock-screen. Thanks to @salvatore373 who did a very good job.
The code-name of this new module is ```flauto```

**Comment by bsutton** [comment]
So how does a new playback system fix a recording problem?

**Comment by Larpoux** [comment]
The recording function is not impacted by this new module.
The recorder code is still same.
What problem are you talking about ?

**Comment by bsutton** [comment]
The work on V2 was to fix a bug with recording under Android X.

**Comment by Larpoux** [comment]
This bug was probably caused by the recorder permission, and is (probably) fixed in the next version.
Unfortunately, this new version has major changes and needs much test/debug before being released.
Sorry for the delay.

**Comment by hyochan** [comment]
@Larpoux Do you mean this `PR` is going to be embedded in #202? By the way, nice code name flauto, I love it 👍 .

**Comment by Larpoux** [comment]
I am busy working on a new PR (flauto) which will include #202 and this #207 .
#202 cannot be released like proposed by @salvatore373 , without fundamental review of the original design/architecture. I have being busy working on flauto those last days, and I hope that I will request a new PR will next week.
Actually there are still several bugs in iOS/Android. @salvatore373 and me are working on them.
Note : It will be a problem for me if I cannot terminate flauto next week, because I have several other important project waiting and I will not have much time for flutter_sound.

**Comment by hyochan** [comment]
@Larpoux Sounds very stunning but I'll be missing you if I don't see you in `flutter_sound` 😢 .
I really appreciate all the hard work you've given so far and I'll do my best to take over all the work you've given to the community.

**Comment by Larpoux** [comment]
Oh NO!. I have some other projects waiting for me, but you will certainly see me again in the future :
- Support the Apple Watch,
- Support OPUS recording on Android
- Try to understand why Flutter Embedded V2 does not work with flauto
- …

I will certainly come back 😃 .
Thank you very much for your nice comment.

**Comment by hyochan** [comment]
Sounds very nice my friend 😄

**Pull request closed without merging**